### PR TITLE
Replace Spans by counters for recvMsg and sendMsg driver events.

### PR DESCRIPTION
**Pull request by neyguvj**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Issue Number: N/A

## What is the new behavior?

-
-
-

### CODE REVIEW

File: driver.go [hunk 1]
```diff
@@ -113,9 +114,11 @@ func driver(cfg *config) trace.Driver {
 		},
 		OnConnNewStream: func(info trace.DriverConnNewStreamStartInfo) func(trace.DriverConnNewStreamDoneInfo) {
 			*info.Context = withTraceID(*info.Context)
-			if cfg.detailer.Details()&trace.DriverConnEvents == 0 {
+			if cfg.detailer.Details()&trace.DriverConnStreamEvents == 0 {
 				return nil
 			}
+			*info.Context = withReceiveMessagesCount(*info.Context, &atomic.Int64{})
```
Comment (asmyasnikov) [inline driver.go]: What do you think about single context creator without external counters?

Comment (asmyasnikov) [reply driver.go]: For example
```
*info.Context = withCounters(*info.Context)
```

Comment (neyguvj) [reply driver.go]: it's a good idea

File: driver.go [hunk 2]
```diff
@@ -391,3 +417,45 @@ func driver(cfg *config) trace.Driver {
...
+func countersFromContext(ctx context.Context) *counters {
+	value, ok := ctx.Value(countersKey{}).(*counters)
+	if !ok {
+		return nil
+	}
+	return value
+}
+
+type counters struct {
+	sent     *atomic.Int64
```
Comment (asmyasnikov) [inline driver.go]:
```suggestion
	sent     atomic.Int64
```

File: driver.go [hunk 3]
```diff
@@ -391,3 +417,45 @@ func driver(cfg *config) trace.Driver {
...
+	received *atomic.Int64
```
Comment (asmyasnikov) [inline driver.go]:
```suggestion
	received atomic.Int64
```

File: driver.go [hunk 4]
```diff
@@ -391,3 +417,45 @@ func driver(cfg *config) trace.Driver {
...
+}
+
+func newCounters() *counters {
+	return &counters{
```
Comment (asmyasnikov) [inline driver.go]:
```suggestion
	return &counters{}
```

File: driver.go [hunk 5]
```diff
@@ -391,3 +417,45 @@ func driver(cfg *config) trace.Driver {
...
+		sent:     &atomic.Int64{},
```
Comment (asmyasnikov) [inline driver.go]:
```suggestion
```

File: driver.go [hunk 6]
```diff
@@ -391,3 +417,45 @@ func driver(cfg *config) trace.Driver {
...
+		received: &atomic.Int64{},
```
Comment (asmyasnikov) [inline driver.go]:
```suggestion
```

File: driver.go [hunk 7]
```diff
@@ -391,3 +417,43 @@ func driver(cfg *config) trace.Driver {
...
+}
+
+type counters struct {
+	sent     *atomic.Int64
+	received *atomic.Int64
+}
+
+func newCounters() *counters {
+	return &counters{}
+	}
```
Comment (asmyasnikov) [inline driver.go]:
```suggestion
```

File: driver.go [hunk 8]
```diff
@@ -391,3 +417,43 @@ func driver(cfg *config) trace.Driver {
...
+	return value
+}
+
+type counters struct {
+	sent     *atomic.Int64
+	received *atomic.Int64
+}
+
+func newCounters() *counters {
```
Comment (asmyasnikov) [inline driver.go]:
```suggestion
```

File: driver.go [hunk 9]
```diff
@@ -391,3 +417,43 @@ func driver(cfg *config) trace.Driver {
...
+	return &counters{}
```
Comment (asmyasnikov) [inline driver.go]:
```suggestion
```

File: driver.go (hunk 7 above)
Comment (asmyasnikov) [inline driver.go]:
```suggestion
```

File: driver.go [hunk 10]
```diff
@@ -391,3 +417,43 @@ func driver(cfg *config) trace.Driver {
...
+	}
+}
```
Comment (asmyasnikov) [inline driver.go:438]:
```suggestion
```

File: driver.go [hunk 11]
```diff
@@ -391,3 +417,43 @@ func driver(cfg *config) trace.Driver {
...
+		return nil
+	}
+	return value
+}
+
+type counters struct {
+	sent     *atomic.Int64
+	received *atomic.Int64
+}
+
```
Comment (asmyasnikov) [inline driver.go]:
```suggestion
```

File: driver.go [hunk 12]
```diff
@@ -391,3 +417,43 @@ func driver(cfg *config) trace.Driver {
 		},
 	}
 }
+
+type countersKey struct{}
+
+func withMessageCounters(ctx context.Context) context.Context {
+	return context.WithValue(ctx, countersKey{}, newCounters())
```
Comment (asmyasnikov) [inline driver.go]:
```suggestion
	return context.WithValue(ctx, countersKey{}, &counters{})
```

File: driver.go [hunk 13]
```diff
@@ -391,3 +417,38 @@ func driver(cfg *config) trace.Driver {
 		},
 	}
 }
+
+type countersKey struct{}
```
Comment (asmyasnikov) [inline driver.go]:
```suggestion
type grpcStreamMsgCountersKey struct{}
```

File: driver.go [hunk 14]
```diff
@@ -391,3 +417,38 @@ func driver(cfg *config) trace.Driver {
...
+
+func withMessageCounters(ctx context.Context) context.Context {
+	return context.WithValue(ctx, countersKey{}, &counters{})
```
Comment (asmyasnikov) [inline driver.go]:
```suggestion
	return context.WithValue(ctx, grpcStreamMsgCountersKey{}, &counters{})
```

File: driver.go [hunk 15]
```diff
@@ -391,3 +417,38 @@ func driver(cfg *config) trace.Driver {
...
+}
+
+func countersFromContext(ctx context.Context) *counters {
+	value, ok := ctx.Value(countersKey{}).(*counters)
```
Comment (asmyasnikov) [inline driver.go]:
```suggestion
	value, ok := ctx.Value(grpcStreamMsgCountersKey{}).(*counters)
```

File: driver.go [hunk 16]
```diff
@@ -391,3 +417,38 @@ func driver(cfg *config) trace.Driver {
...
+
+func countersFromContext(ctx context.Context) *counters {
+	value, ok := ctx.Value(grpcStreamMsgCountersKey{}).(*counters)
+	if !ok {
+		return nil
+	}
+	return value
+}
+
+type counters struct {
```
Comment (asmyasnikov) [inline driver.go]:
```suggestion
type grpcStreamMsgCounters struct {
```

File: driver.go [hunk 17]
```diff
@@ -391,3 +417,38 @@ func driver(cfg *config) trace.Driver {
...
+	sent     atomic.Int64
+	received atomic.Int64
+}
+
+func (c *counters) updateSentMessages() {
```
Comment (asmyasnikov) [inline driver.go]:
```suggestion
func (c *grpcStreamMsgCounters) updateSentMessages() {
```

File: driver.go [hunk 18]
```diff
@@ -391,3 +417,38 @@ func driver(cfg *config) trace.Driver {
...
+	c.sent.Add(1)
+}
+
+func (c *counters) updateReceivedMessages() {
```
Comment (asmyasnikov) [inline driver.go]:
```suggestion
func (c *grpcStreamMsgCounters) updateReceivedMessages() {
```

File: driver.go [hunk 19]
```diff
@@ -391,3 +417,38 @@ func driver(cfg *config) trace.Driver {
...
+	c.received.Add(1)
+}
+
+func (c *counters) sentMessages() int64 {
```
Comment (asmyasnikov) [inline driver.go]:
```suggestion
func (c *grpcStreamMsgCounters) sentMessages() int64 {
```

File: driver.go [hunk 20]
```diff
@@ -391,3 +417,38 @@ func driver(cfg *config) trace.Driver {
...
+	return c.sent.Load()
+}
+
+func (c *counters) receivedMessages() int64 {
```
Comment (asmyasnikov) [inline driver.go]:
```suggestion
func (c *grpcStreamMsgCounters) receivedMessages() int64 {
```

File: driver.go [hunk 21]
```diff
@@ -391,3 +417,38 @@ func driver(cfg *config) trace.Driver {
 		},
 	}
 }
+
+type grpcStreamMsgCountersKey struct{}
+
+func withMessageCounters(ctx context.Context) context.Context {
```
Comment (asmyasnikov) [inline driver.go]:
```suggestion
func withMessageCounters(ctx context.Context) context.Context {
```
```suggestion
func withGrpcStreamMsgCounters(ctx context.Context) context.Context {
```

File: driver.go [hunk 22]
```diff
@@ -113,9 +114,10 @@ func driver(cfg *config) trace.Driver {
 		},
 		OnConnNewStream: func(info trace.DriverConnNewStreamStartInfo) func(trace.DriverConnNewStreamDoneInfo) {
 			*info.Context = withTraceID(*info.Context)
-			if cfg.detailer.Details()&trace.DriverConnEvents == 0 {
+			if cfg.detailer.Details()&trace.DriverConnStreamEvents == 0 {
 				return nil
 			}
+			*info.Context = withMessageCounters(*info.Context)
```
Comment (asmyasnikov) [inline driver.go]:
```suggestion
			*info.Context = withGrpcStreamMsgCounters(*info.Context)
```

File: driver.go [hunk 23]
```diff
@@ -391,3 +417,38 @@ func driver(cfg *config) trace.Driver {
...
 	}
 }
+
+type grpcStreamMsgCountersKey struct{}
+
+func withGrpcStreamMsgCounters(ctx context.Context) context.Context {
+	return context.WithValue(ctx, grpcStreamMsgCountersKey{}, &counters{})
+}
+
+func countersFromContext(ctx context.Context) *counters {
```
Comment (asmyasnikov) [inline driver.go]:
```suggestion
func grpcStreamMsgCountersFromContext(ctx context.Context) *counters {
```

File: driver.go [hunk 24]
```diff
@@ -173,6 +179,26 @@ func driver(cfg *config) trace.Driver {
...
+		OnConnStreamFinish: func(info trace.DriverConnStreamFinishInfo) {
+			if cfg.detailer.Details()&trace.DriverConnStreamEvents == 0 {
+				return
+			}
+			start := childSpanWithReplaceCtx(
+				cfg.tracer,
+				&info.Context,
+				info.Call.FunctionID(),
+			)
+			counters := countersFromContext(info.Context)
```
Comment (asmyasnikov) [inline driver.go]:
```suggestion
			counters := grpcStreamMsgCountersFromContext(info.Context)
```

File: driver.go [hunk 25]
```diff
@@ -131,37 +133,41 @@ func driver(cfg *config) trace.Driver {
...
+			if cfg.detailer.Details()&trace.DriverConnStreamEvents == 0 {
 				return nil
 			}
-			start := childSpanWithReplaceCtx(
-				cfg.tracer,
-				info.Context,
-				info.Call.FunctionID(),
-			)
+			start := otelTrace.SpanFromContext(*info.Context)
+			counters := countersFromContext(*info.Context)
```
Comment (asmyasnikov) [inline driver.go]:
```suggestion
			counters := grpcStreamMsgCountersFromContext(*info.Context)
```

File: driver.go [hunk 26]
```diff
@@ -131,37 +133,41 @@ func driver(cfg *config) trace.Driver {
...
```
Comment (asmyasnikov) [inline driver.go]:
```suggestion
			counters := grpcStreamMsgCountersFromContext(*info.Context)
```

File: driver.go [hunk 27]
```diff
@@ -173,6 +179,26 @@ func driver(cfg *config) trace.Driver {
...
+				return
+			}
+			start := childSpanWithReplaceCtx(
+				cfg.tracer,
+				&info.Context,
+				info.Call.FunctionID(),
+			)
+			counters := grpcStreamMsgCountersFromContext(info.Context)
+
+			var attributes []attribute.KeyValue
```
Comment (asmyasnikov) [inline driver.go]:
```suggestion
			attributes := make([]attribute.KeyValue, 0, 2)
```

File: driver.go [hunk 28]
```diff
@@ -173,6 +179,26 @@ func driver(cfg *config) trace.Driver {
...
+			if counters != nil {
+				attributes = []attribute.KeyValue{
```
Comment (asmyasnikov) [inline driver.go]:
```suggestion
				attributes = append(attributes,
```

File: driver.go [hunk 29]
```diff
@@ -173,6 +179,26 @@ func driver(cfg *config) trace.Driver {
...
+					attribute.Int64("received_messages", counters.receivedMessages()),
+					attribute.Int64("sent_messages", counters.sentMessages()),
+				}
```
Comment (asmyasnikov) [inline driver.go]:
```suggestion
				)
```

**Pull request merged**